File: src/backend/entities/src/main/java/com/activepieces/common/pagination/SeekPage.java
package com.activepieces.common.pagination;

import lombok.Getter;
import lombok.NoArgsConstructor;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Function;
import java.util.stream.Collectors;

@Getter
@NoArgsConstructor
public class SeekPage<T> {

    private Cursor previous;
    private Cursor next;
    private List<T> data;

    public SeekPage(List<T> data,  Cursor previous,Cursor next) {
        this.previous = previous;
        this.next = next;
        this.data = data;
    }


    public <S> SeekPage<S> convert(Function<T, S> function) {
        return new SeekPage<>(
                data.stream()
                        .map(function)
                        .collect(Collectors.toCollection(CopyOnWriteArrayList::new)),
                previous, next);
    }

}

File: packages/shared/src/lib/core/common/seek-page.ts
import { z } from 'zod'
import { Nullable } from './base-model'

export type Cursor = string | null

export type SeekPage<T> = {
    next: Cursor
    previous: Cursor
    data: T[]
}

export const SeekPage = (t: z.ZodType): z.ZodType => z.object({
    data: z.array(t),
    next: Nullable(z.string()),
    previous: Nullable(z.string()),
})
